pkg/workflows/wasm/runner_wasip1.go

```diff
@@ -23,107 +23,138 @@ func log(respptr unsafe.Pointer, respptrlen int32)
+				emitFn:  emitFn,
 			}
 		},
 		args: os.Args,
 	}
 }
 
+// sendResponseFn implements sendResponse for import into WASM.
+func sendResponseFn(response *wasmpb.Response) {
+	pb, err := proto.Marshal(response)
+	if err != nil {
+		// We somehow couldn't marshal the response, so let's
+		// exit with a special error code letting the host know
+		// what happened.
+		os.Exit(CodeInvalidResponse)
+	}
+
+	// unknownID will only be set when we've failed to parse
+	// the request. Like before, let's bubble this up.
+	if response.Id == unknownID {
+		os.Exit(CodeInvalidRequest)
+	}
+
+	ptr, ptrlen := bufferToPointerLen(pb)
+	errno := sendResponse(ptr, ptrlen)
+	if errno != 0 {
+		os.Exit(CodeHostErr)
+	}
+
+	code := CodeSuccess
+	if response.ErrMsg != "" {
+		code = CodeRunnerErr
+	}
+
+	os.Exit(code)
+}
+
+// createFetchFn injects dependencies and creates a fetch function that can be used by the WASM
+// binary.
+func createFetchFn(
+	sdkConfig *RuntimeConfig,
+	l logger.Logger,
+) func(sdk.FetchRequest) (sdk.FetchResponse, error) {
+	fetchFn := func(req sdk.FetchRequest) (sdk.FetchResponse, error) {
+		headerspb, err := values.NewMap(req.Headers)
+		if err != nil {
+			os.Exit(CodeInvalidRequest)
+		}
+
+		b, err := proto.Marshal(&wasmpb.FetchRequest{
+			Url:       req.URL,
+			Method:    req.Method,
+			Headers:   values.ProtoMap(headerspb),
+			Body:      req.Body,
+			TimeoutMs: req.TimeoutMs,
+		})
+		if err != nil {
+			os.Exit(CodeInvalidRequest)
+		}
+		reqptr, reqptrlen := bufferToPointerLen(b)
+
+		respBuffer := make([]byte, sdkConfig.MaxFetchResponseSizeBytes)
+		respptr, _ := bufferToPointerLen(respBuffer)
+
+		resplenBuffer := make([]byte, uint32Size)
+		resplenptr, _ := bufferToPointerLen(resplenBuffer)
+
+		errno := fetch(respptr, resplenptr, reqptr, reqptrlen)
+		if errno != 0 {
+			os.Exit(CodeRunnerErr)
+		}
+
+		responseSize := binary.LittleEndian.Uint32(resplenBuffer)
+		response := &wasmpb.FetchResponse{}
+		err = proto.Unmarshal(respBuffer[:responseSize], response)
+		if err != nil {
+			l.Errorw("failed to unmarshal fetch response", "error", err.Error())
+			os.Exit(CodeInvalidResponse)
+		}
+
+		fields := response.Headers.GetFields()
+		headersResp := make(map[string]any, len(fields))
+		for k, v := range fields {
+			headersResp[k] = v
+		}
+
+		if response.ErrorMessage != "" {
+			return sdk.FetchResponse{}, errors.New(response.ErrorMessage)
+		}
+
+		return sdk.FetchResponse{
+			Success:    response.Success,
+			StatusCode: uint8(response.StatusCode),
+			Headers:    headersResp,
+			Body:       response.Body,
+		}, nil
+	}
+	return fetchFn
+}
+
+func emitFn(msg string, labels map[string]any) error {
+	vm, err := values.NewMap(labels)
+	if err != nil {
+		return err
+	}
+
+	b, err := proto.Marshal(&wasmpb.CustomEmitMessage{
+		Message: msg,
+		Labels:  values.ProtoMap(vm),
+	})
+
+	ptr, ptrlen := bufferToPointerLen(b)
+	errno := emit(ptr, ptrlen)
+	if errno != 0 {
+		os.Exit(CodeRunnerErr)
```

Comment:
@MStreet3 You can return an error here rather than exiting: let the caller handle this and choose what to do; if they want to exit, they can bubble up the error.